<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset such as .csv, .json, .xlsx  etc. The dataset can be stored in different places on your local machine or sometimes online.<br>


In [1]:
import pandas as pd

df = pd.read_csv('dataset/uncleaned_auto_data.csv') 
print(" 5 rows of the dataframe") 
df.head(5)

 5 rows of the dataframe


Unnamed: 0.1  Unnamed: 0  symboling  normalized-losses         make  \
0             0           0          3                122  alfa-romero   
1             1           1          3                122  alfa-romero   
2             2           2          1                122  alfa-romero   
3             3           3          2                164         audi   
4             4           4          2                164         audi   

  aspiration num-of-doors   body-style drive-wheels engine-location  ...  \
0        std          two  convertible          rwd           front  ...   
1        std          two  convertible          rwd           front  ...   
2        std          two    hatchback          rwd           front  ...   
3        std         four        sedan          fwd           front  ...   
4        std         four        sedan          4wd           front  ...   

   compression-ratio  horsepower  peak-rpm  city-mpg  highway-mpg    price  \
0                9.0       111.0    5000.0        21           27  13495.0   
1                9.0       111.0    5000.0        21           27  16500.0   
2                9.0       154.0    5000.0        19           26  16500.0   
3               10.0       102.0    5500.0        24           30  13950.0   
4                8.0       115.0    5500.0        18           22  17450.0   

  city-L/100km  horsepower-binned diesel  gas  
0    11.190476             Medium      0    1  
1    11.190476             Medium      0    1  
2    12.368421             Medium      0    1  
3     9.791667             Medium      0    1  
4    13.055556             Medium      0    1  

[5 rows x 31 columns]

Drop the missing values along the column "price"  

In [9]:
df2 = df.drop(columns=['Unnamed: 0.1', 'Unnamed: 0'], axis=0)

In [10]:
print("COLUMN NAMES")
print(df2.columns)

COLUMN NAMES
Index(['symboling', 'normalized-losses', 'make', 'aspiration', 'num-of-doors',
       'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length',
       'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders',
       'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio',
       'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price',
       'city-L/100km', 'horsepower-binned', 'diesel', 'gas'],
      dtype='object')


In [ ]:
df2.to_csv('dataset/final_data.csv' , index=False)

<h2>Data Types</h2>
<p>
Data has a variety of types.</b>The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good to know the data type of each column. In Pandas:
</p>

In [11]:
df2.dtypes


symboling              int64
normalized-losses      int64
make                  object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
city-L/100km         float64
horsepower-binned     object
diesel                 int64
gas                    int64
dtype: object

It returns a Series with the data type of each column.

<p>
It is clear to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>

In [5]:
df2.describe()

Unnamed: 0.1  Unnamed: 0   symboling  normalized-losses  wheel-base  \
count    201.000000  201.000000  201.000000          201.00000  201.000000   
mean     100.000000  100.000000    0.840796          122.00000   98.797015   
std       58.167861   58.167861    1.254802           31.99625    6.066366   
min        0.000000    0.000000   -2.000000           65.00000   86.600000   
25%       50.000000   50.000000    0.000000          101.00000   94.500000   
50%      100.000000  100.000000    1.000000          122.00000   97.000000   
75%      150.000000  150.000000    2.000000          137.00000  102.400000   
max      200.000000  200.000000    3.000000          256.00000  120.900000   

           length       width      height  curb-weight  engine-size  ...  \
count  201.000000  201.000000  201.000000   201.000000   201.000000  ...   
mean     0.837102    0.915126   53.766667  2555.666667   126.875622  ...   
std      0.059213    0.029187    2.447822   517.296727    41.546834  ...   
min      0.678039    0.837500   47.800000  1488.000000    61.000000  ...   
25%      0.801538    0.890278   52.000000  2169.000000    98.000000  ...   
50%      0.832292    0.909722   54.100000  2414.000000   120.000000  ...   
75%      0.881788    0.925000   55.500000  2926.000000   141.000000  ...   
max      1.000000    1.000000   59.800000  4066.000000   326.000000  ...   

           stroke  compression-ratio  horsepower     peak-rpm    city-mpg  \
count  197.000000         201.000000  201.000000   201.000000  201.000000   
mean     3.256904          10.164279  103.405534  5117.665368   25.179104   
std      0.319256           4.004965   37.365700   478.113805    6.423220   
min      2.070000           7.000000   48.000000  4150.000000   13.000000   
25%      3.110000           8.600000   70.000000  4800.000000   19.000000   
50%      3.290000           9.000000   95.000000  5125.369458   24.000000   
75%      3.410000           9.400000  116.000000  5500.000000   30.000000   
max      4.170000          23.000000  262.000000  6600.000000   49.000000   

       highway-mpg         price  city-L/100km      diesel         gas  
count   201.000000    201.000000    201.000000  201.000000  201.000000  
mean     30.686567  13207.129353      9.944145    0.099502    0.900498  
std       6.815150   7947.066342      2.534599    0.300083    0.300083  
min      16.000000   5118.000000      4.795918    0.000000    0.000000  
25%      25.000000   7775.000000      7.833333    0.000000    1.000000  
50%      30.000000  10295.000000      9.791667    0.000000    1.000000  
75%      34.000000  16500.000000     12.368421    0.000000    1.000000  
max      54.000000  45400.000000     18.076923    1.000000    1.000000  

[8 rows x 21 columns]

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>
For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.


In [12]:
# describe all the columns in "df" 
df2.describe(include = "all")

symboling  normalized-losses    make aspiration num-of-doors  \
count   201.000000          201.00000     201        201          201   
unique         NaN                NaN      22          2            2   
top            NaN                NaN  toyota        std         four   
freq           NaN                NaN      32        165          115   
mean      0.840796          122.00000     NaN        NaN          NaN   
std       1.254802           31.99625     NaN        NaN          NaN   
min      -2.000000           65.00000     NaN        NaN          NaN   
25%       0.000000          101.00000     NaN        NaN          NaN   
50%       1.000000          122.00000     NaN        NaN          NaN   
75%       2.000000          137.00000     NaN        NaN          NaN   
max       3.000000          256.00000     NaN        NaN          NaN   

       body-style drive-wheels engine-location  wheel-base      length  ...  \
count         201          201             201  201.000000  201.000000  ...   
unique          5            3               2         NaN         NaN  ...   
top         sedan          fwd           front         NaN         NaN  ...   
freq           94          118             198         NaN         NaN  ...   
mean          NaN          NaN             NaN   98.797015    0.837102  ...   
std           NaN          NaN             NaN    6.066366    0.059213  ...   
min           NaN          NaN             NaN   86.600000    0.678039  ...   
25%           NaN          NaN             NaN   94.500000    0.801538  ...   
50%           NaN          NaN             NaN   97.000000    0.832292  ...   
75%           NaN          NaN             NaN  102.400000    0.881788  ...   
max           NaN          NaN             NaN  120.900000    1.000000  ...   

        compression-ratio  horsepower     peak-rpm    city-mpg highway-mpg  \
count          201.000000  201.000000   201.000000  201.000000  201.000000   
unique                NaN         NaN          NaN         NaN         NaN   
top                   NaN         NaN          NaN         NaN         NaN   
freq                  NaN         NaN          NaN         NaN         NaN   
mean            10.164279  103.405534  5117.665368   25.179104   30.686567   
std              4.004965   37.365700   478.113805    6.423220    6.815150   
min              7.000000   48.000000  4150.000000   13.000000   16.000000   
25%              8.600000   70.000000  4800.000000   19.000000   25.000000   
50%              9.000000   95.000000  5125.369458   24.000000   30.000000   
75%              9.400000  116.000000  5500.000000   30.000000   34.000000   
max             23.000000  262.000000  6600.000000   49.000000   54.000000   

               price city-L/100km  horsepower-binned      diesel         gas  
count     201.000000   201.000000                200  201.000000  201.000000  
unique           NaN          NaN                  3         NaN         NaN  
top              NaN          NaN                Low         NaN         NaN  
freq             NaN          NaN                115         NaN         NaN  
mean    13207.129353     9.944145                NaN    0.099502    0.900498  
std      7947.066342     2.534599                NaN    0.300083    0.300083  
min      5118.000000     4.795918                NaN    0.000000    0.000000  
25%      7775.000000     7.833333                NaN    0.000000    1.000000  
50%     10295.000000     9.791667                NaN    0.000000    1.000000  
75%     16500.000000    12.368421                NaN    0.000000    1.000000  
max     45400.000000    18.076923                NaN    1.000000    1.000000  

[11 rows x 29 columns]

<p>
Now, it provides the statistical summary of all the columns, including object-typed attributes.<br>
We can now see how many unique values, which is the top value and the frequency of top value in the object-typed columns.<br>
Some values in the table above show as "NaN", this is because those numbers are not available regarding a particular column type.<br>
</p>

In [15]:
df2[['length','compression-ratio']].describe()

length  compression-ratio
count  201.000000         201.000000
mean     0.837102          10.164279
std      0.059213           4.004965
min      0.678039           7.000000
25%      0.801538           8.600000
50%      0.832292           9.000000
75%      0.881788           9.400000
max      1.000000          23.000000

In [14]:
df2.info

<bound method DataFrame.info of      symboling  normalized-losses         make aspiration num-of-doors  \
0            3                122  alfa-romero        std          two   
1            3                122  alfa-romero        std          two   
2            1                122  alfa-romero        std          two   
3            2                164         audi        std         four   
4            2                164         audi        std         four   
..         ...                ...          ...        ...          ...   
196         -1                 95        volvo        std         four   
197         -1                 95        volvo      turbo         four   
198         -1                 95        volvo        std         four   
199         -1                 95        volvo      turbo         four   
200         -1                 95        volvo      turbo         four   

      body-style drive-wheels engine-location  wheel-base    length  ...  \
0  

<p>
It shows us that the whole data frame has 205 rows and 26 columns in total.
</p>